hive-metastore/src/main/java/org/apache/iceberg/hive/CachedClientPool.java

```diff
@@ -0,0 +1,84 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.hive;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.apache.iceberg.CatalogProperties;
+import org.apache.iceberg.ClientPool;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.util.PropertyUtil;
+import org.apache.thrift.TException;
+
+public class CachedClientPool implements ClientPool<IMetaStoreClient, TException> {
+
+  private static Cache<String, HiveClientPool> clientPoolCache;
+
+  private final Configuration conf;
+  private final String metastoreUri;
+  private final int clientPoolSize;
+  private final long evictionInterval;
+
+  CachedClientPool(Configuration conf, Map<String, String> properties) {
+    this.conf = conf;
+    this.metastoreUri = conf.get(HiveConf.ConfVars.METASTOREURIS.varname, "");
+    this.clientPoolSize = PropertyUtil.propertyAsInt(properties,
+            CatalogProperties.CLIENT_POOL_SIZE,
+            CatalogProperties.CLIENT_POOL_SIZE_DEFAULT);
+    this.evictionInterval = PropertyUtil.propertyAsLong(properties,
+            CatalogProperties.CLIENT_POOL_CACHE_EVICTION_INTERVAL_MS,
+            CatalogProperties.CLIENT_POOL_CACHE_EVICTION_INTERVAL_MS_DEFAULT);
+    init();
+  }
+
+  @VisibleForTesting
+  HiveClientPool clientPool() {
+    return clientPoolCache.get(metastoreUri, k -> new HiveClientPool(clientPoolSize, conf));
+  }
+
+  private synchronized void init() {
+    if (clientPoolCache == null) {
+      clientPoolCache = Caffeine.newBuilder().expireAfterAccess(evictionInterval, TimeUnit.MILLISECONDS)
+              .removalListener((key, value, cause) -> ((HiveClientPool) value).close())
+              .build();
+    }
+  }
+
+  @VisibleForTesting
+  static Cache<String, HiveClientPool> clientPoolCache() {
+    return clientPoolCache;
+  }
+
+  @Override
+  public <R> R run(Action<R, IMetaStoreClient, TException> action) throws TException, InterruptedException {
+    return clientPool().run(action);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the resource returned by the *clientPool* method in a try-finally block. Execution paths that do not contain closure statements were detected.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).